include/ecdaa/prng.h

```diff
@@ -0,0 +1,102 @@
+/******************************************************************************
+ *
+ * Copyright 2017 Xaptum, Inc.
+ * 
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ * 
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License
+ *
+ *****************************************************************************/
+
+#ifndef ECDAA_RANDOM_H
+#define ECDAA_RANDOM_H
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <amcl/randapi.h>
+
+#include <stddef.h>
+
+#define AMCL_SEED_SIZE 128  // As recommended in AMCL's "rand.c"
+
+enum ecdaa_prng_initialized {
+    ECDAA_PRNG_INITIALIZED_YES,
+    ECDAA_PRNG_INITIALIZED_NO
+};
+
+/*
+ * Wrapper around AMCL's pseudo-random-number-generator.
+ *
+ * For the safe operation of this library,
+ * all `ecdaa_prng`s MUST first be properly seeded
+ * via a call to `ecdaa_prng_init`
+ * of `ecdaa_prng_init_custom` before first use.
+ */
+struct ecdaa_prng {
+    enum ecdaa_prng_initialized initialized;
+    csprng impl;
+};
+
+#ifndef DISABLE_LIBSODIUM_RNG_SEED_FUNCTION
+
+/*
+ * Properly seed a `ecdaa_prng`.
+ *
+ * MUST be called on a `ecdaa_prng` before first use.
+ *
+ * No dynamic memory allocation is performed.
+ *
+ * Uses Libsodium's `randombytes_buf` as the source of
+ * a cryptographically-strong random seed.
+ * Requires the CMake option `DISABLE_LIBSODIUM_RNG_SEED_FUNCTION=OFF`.
+ *
+ * Returns:
+ * 0 on success
+ * -1 if unable to obtain seed
+ * -2 if entropy is insufficient
+ */
+int ecdaa_prng_init(struct ecdaa_prng *prng_in);
+
+#endif  // DISABLE_LIBSODIUM_RNG_SEED_FUNCTION
+
+/*
+ * Type of function used to obtain a cryptographically-secure seed.
+ *
+ * Must be able to provide AMCL_SEED_SIZE crpytographically-strong bytes in a single call.
+ *
+ * Returns:
+ * 0 on success
+ * non-zero on failure
+ */
+typedef int (*ecdaa_prng_custom_seed_function_type)(void * const buf, const size_t size);
+
+/*
+ * Properly seed a `ecdaa_prng`.
+ *
+ * MUST be called on a `ecdaa_prng` before first use.
+ *
+ * No dynamic memory allocation is performed.
+ *
+ * Returns:
+ * 0 on success
+ * return value of `custom_func`, if non-zero
+ */
+int ecdaa_prng_init_custom(struct ecdaa_prng *prng_in, ecdaa_prng_custom_seed_function_type custom_func);
```

Comment:
Maybe just take the seed directly, rather than a seed function?  My gut feeling is that will be the easiest API for users.

Reply:
I want to make sure we actually get `AMCL_SEED_SIZE` bytes of seed. I can imagine someone thinking "well, <x> bytes is an appropriate seed size, so that's what I'll pass in" without realizing we're requesting a specific size, then we try to read more bytes than they gave us.

We can document the requirement by taking the seed as `char seed[AMCL_SEED_SIZE]`, but we have no way of determining if that's what is actually passed in.

Maybe I'm being anal, and this is a caveat emptor situation. And, yea, I can see that having to pass in a function with a specific signature can be a pain.

Reply:
No, after thinking about it more, I think you're right. I think I'm trying to hard to tell the user what to do, and making the API hard to use.

I agree now.

Reply:
Can we take both the pointer and size as arguments, and return an error if the size is less than `AMCL_SEED_SIZE`?

Reply:
:+1: